### PR TITLE
feat: separate toxic players from new players in matchmaking

### PR DESCRIPTION
**Pull request by thesprockee**
## Summary

- Prevents players with suspension history from being matched with new players (games_played < NewPlayerMaxGames threshold)
- Enforcers and global operators are **exempt** -- their suspension records may come from admin work, not toxicity
- Only blocks candidates containing BOTH a new player AND a non-exempt player with `has_suspension_history = "true"`
- Veteran-only matches are unaffected -- this is a new-player shield, not a general punishment

## Implementation

- **`server/evr_toxic_separation.go`** -- `HasSuspensionHistory()`, `CandidateContainsToxicNewPlayerMix()`, `FilterToxicNewPlayerCandidates()` core logic
- **`server/evr_lobby_parameters.go`** -- Load enforcement journal at ticket creation, set `has_suspension_history` string property; moderators (enforcers/operators) always get `"false"`
- **`server/evr_global_settings.go`** -- `EnableToxicSeparation` setting (`*bool`, default true) with `ToxicSeparationEnabled()` helper
- **`server/evr_matchmaker_process.go`** -- Filter toxic-new-player candidates in `processPotentialMatches` before prediction

## Depends on

- `feat/games-played-tracking` (PR 1: `games_played` property + `IsNewPlayer` helper)

## Test plan

- [x] new player + toxic player -> candidate rejected
- [x] new player + clean player -> candidate accepted
- [x] veteran + toxic player -> candidate accepted (no new player = no shield)
- [x] new player + enforcer with suspensions -> accepted (enforcer exempt)
- [x] disabled setting (threshold 0) passes all candidates
- [x] missing `has_suspension_history` property treated as clean
- [x] `ToxicSeparationEnabled()` defaults to true, respects explicit true/false
- [x] `FixDefaultServiceSettings` sets default
- [x] `HasSuspensionHistory()` helper edge cases
- [x] `go build ./server/...` passes
- [x] `go test ./server/... -run "TestToxic|TestSeparation|TestHasSuspension|TestCandidateContains|TestFilter" -v` -- all 17 tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by thesprockee** [comment]
Addressed all three Copilot review comments in f76585b3:

**Comment 3 (multiple `ServiceSettings()` calls):** Fixed. `ServiceSettings()` is now loaded once at the top of `processPotentialMatches` and reused throughout the function, preventing inconsistent config snapshots during hot updates.

**Comment 2 (unnecessary journal read when `NewPlayerMaxGames <= 0`):** Already addressed in the original commit — the journal read condition includes `globalSettings.NewPlayerMaxGames > 0`.

**Comment 1 (party bypass):** Documented as a known limitation with a comment in `addTicket`. For party tickets, `has_suspension_history` and `games_played` reflect only the leader, not all party members. The fix (aggregating across party members) requires changes to party ticket property computation and is deferred.

Build compiles, all related tests pass. CI failure is a pre-existing build infrastructure issue in `internal/gopher-lua/parse/parser.go.y` (C yacc trying to compile Go code) — unrelated to this PR.

File: server/evr_matchmaker_process.go [hunk 1]
```diff
@@ -17,6 +17,12 @@ func (m *SkillBasedMatchmaker) processPotentialMatches(logger runtime.Logger, en
...
 
+	// Filter out candidates where a new player would be matched with a
+	// player who has suspension history (toxic player separation).
+	if settings := ServiceSettings(); settings != nil && settings.Matchmaking.ToxicSeparationEnabled() {
+		filterCounts["toxic_separation"] = FilterToxicNewPlayerCandidates(candidates, settings.Matchmaking.NewPlayerMaxGames)
+	}
+
 	config := PredictionConfig{}
 	if settings := ServiceSettings(); settings != nil {
 		mu := settings.SkillRating.Defaults.Mu
```
[the remark this answers is not in the document]
Comment (thesprockee) [reply server/evr_matchmaker_process.go:28]: The toxic separation `ServiceSettings()` call added by this PR was removed in a subsequent refactor (the archetype detection feature in fa7d9a2..d147511 restructured how new-player detection works, removing `FilterToxicNewPlayerCandidates` entirely).

The remaining three `ServiceSettings()` calls (prediction config, wait-time threshold, and reservation logic) predate this PR and are a pre-existing concern. Since `ServiceSettings()` is an atomic pointer load, the correctness risk from version skew mid-function is real but low-probability. I'll address this in a separate follow-up that caches the settings at the top of `processPotentialMatches`.

File: server/evr_matchmaker_process.go [hunk 2]
```diff
@@ -17,6 +17,12 @@ func (m *SkillBasedMatchmaker) processPotentialMatches(logger runtime.Logger, en
 	// Filter out players who are too far away from each other
 	filterCounts["max_rtt"] = m.filterWithinMaxRTT(candidates)
 
+	// Filter out candidates where a new player would be matched with a
+	// player who has suspension history (toxic player separation).
+	if settings := ServiceSettings(); settings != nil && settings.Matchmaking.ToxicSeparationEnabled() {
+		filterCounts["toxic_separation"] = FilterToxicNewPlayerCandidates(candidates, settings.Matchmaking.NewPlayerMaxGames)
+	}
```
[the remark this answers is not in the document]
Comment (thesprockee) [reply server/evr_matchmaker_process.go:24]: Acknowledged — this is a known limitation. Party tickets reflect the leader's properties because `PartyHandler.MatchmakerAdd` submits a single set of `stringProperties`/`numericProperties` for the entire party. The archetype detection feature (fa7d9a2) that replaced this PR's approach also has the same limitation (archetype is detected for the leader only).

Correctly aggregating party-wide properties (e.g., `has_suspension_history = true` if ANY member has history, `games_played = MIN(members)`) requires changes to ticket construction in `addTicket` and is tracked as a separate enhancement.

File: server/evr_lobby_parameters.go [hunk 3]
```diff
@@ -334,6 +335,38 @@ func NewLobbyParametersFromRequest(ctx context.Context, logger *zap.Logger, nk r
...
+						hasSuspensionHistory = true
+						break
+					}
+				}
+				if hasSuspensionHistory {
+					break
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (thesprockee) [reply server/evr_lobby_parameters.go:368]: This code was removed in a subsequent refactor. The archetype detection feature (fa7d9a2..d147511) replaced the journal-based suspension history approach entirely — the enforcement journal is no longer loaded during matchmaking ticket creation. New-player detection now uses `games_played` from the GamesPlayed leaderboard and archetype classification from rolling match stats, with no storage reads for enforcement data.

The concern about unnecessary journal reads when `NewPlayerMaxGames <= 0` is moot since the journal read no longer exists.
